In [ ]:
#costma

In [118]:
import csv

with open('kaggle2021-short.csv') as f:
    reader = csv.reader(f, delimiter=",")
    kaggle_data = list(reader)

column_names = kaggle_data[0]
survey_responses = kaggle_data[1:]


In [119]:
# Iterate over the indices so that we can update all of the data
num_rows = len(survey_responses)
for i in range(num_rows):

    # experience_coding
    survey_responses[i][0] = float(survey_responses[i][0]) 
    
    # python_user
    if survey_responses[i][1] == "TRUE":
        survey_responses[i][1] = True
    else:
        survey_responses[i][1] = False
    
    # r_user
    if survey_responses[i][2] == "TRUE":
        survey_responses[i][2] = True
    else:
        survey_responses[i][2] = False

    # sql_user
    if survey_responses[i][3] == "TRUE":
        survey_responses[i][3] = True
    else:
        survey_responses[i][3] = False

    # most_used
    if survey_responses[i][4] == "None":
        survey_responses[i][4] = None
    else:
        survey_responses[i][4] = survey_responses[i][4]


    # compensation
    survey_responses[i][5] = int(survey_responses[i][5])    

In [120]:
num_rows = len(survey_responses)
python_user_count = 0
r_user_count = 0
sql_user_count = 0

for i in range(num_rows):

    # Detect if python_user column is True
    if survey_responses[i][1]:
        python_user_count = python_user_count + 1
    
    # Detect if r_user column is True
    if survey_responses[i][2]:
        r_user_count = r_user_count + 1

    # Detect if sql_user column is True
    if survey_responses[i][3]:
        sql_user_count = sql_user_count + 1

print("Number of Python users: " + str(python_user_count))
print("Number of R users: " + str(r_user_count))
print("Number of SQL users: " + str(sql_user_count))

print("Proportion of Python users: " + str(python_user_count / num_rows))
print("Proportion of R users: " + str(r_user_count  / num_rows))
print("Proportion of SQL users: " + str(sql_user_count  / num_rows))

Number of Python users: 21860
Number of R users: 5335
Number of SQL users: 10757
Proportion of Python users: 0.8416432449081739
Proportion of R users: 0.20540561352173412
Proportion of SQL users: 0.4141608593539445


In [121]:
experience_coding = []
compensation = []

for row in survey_responses:
    experience_coding.append(row[0])
    compensation.append(row[5])
    
maximum_experience = max(experience_coding)
minimum_experience = min(experience_coding) 
average_experience = round((sum(experience_coding)) / (len(experience_coding)),2)

maximum_experience_comp = max(compensation)
minimum_experience_comp = min(compensation) 
average_experience_comp = round((sum(compensation)) / (len(compensation)),2)

print("The longest experience:", maximum_experience)
print("The shortest experience:", minimum_experience)
print("Average experience of all users:",average_experience)
print("The highest salary:",maximum_experience_comp,"$")
print("The smallest salary:",minimum_experience_comp,"$")
print("Average salaries of all users:",average_experience_comp,"$")

The longest experience: 30.0
The shortest experience: 0.0
Average experience of all users: 5.3
The highest salary: 1492951 $
The smallest salary: 0 $
Average salaries of all users: 53252.82 $


In [122]:

for i in range(num_rows):

    if survey_responses[i][0] < 5:
        survey_responses[i].append("<5 Years")
    
    elif survey_responses[i][0] >= 5 and survey_responses[i][0] < 10:
        survey_responses[i].append("5-10 Years")

    elif survey_responses[i][0] >= 10 and survey_responses[i][0] < 15:
        survey_responses[i].append("10-15 Years")
    
    elif survey_responses[i][0] >= 15 and survey_responses[i][0] < 20:
        survey_responses[i].append("15-20 Years")

    elif survey_responses[i][0] >= 20 and survey_responses[i][0] < 25:
        survey_responses[i].append("20-25 Years")
    
    else:
        survey_responses[i].append("25+ Years")
print(survey_responses[1:5])

[[12.3, True, True, True, 'Scikit-learn', 236889, '10-15 Years'], [2.2, True, False, False, None, 74321, '<5 Years'], [2.7, False, False, True, None, 62593, '<5 Years'], [1.2, True, False, False, 'Scikit-learn', 36288, '<5 Years']]


In [ ]:
List_below_5 = []
List_5_10 = []
List_10_15 = []
List_15_20 = []
List_20_25 = []
List_above_25 = []

for i in range(num_rows):

    if survey_responses[i][6] == "<5 Years":
        List_below_5.append(survey_responses[i][5])
        
        
    
    elif survey_responses[i][6] == "5-10 Years":
        List_5_10.append(survey_responses[i][5])

    elif survey_responses[i][6] == "10-15 Years":
        List_10_15.append(survey_responses[i][5])
    
    elif survey_responses[i][6] == "15-20 Years":
        List_15_20.append(survey_responses[i][5])

    elif survey_responses[i][6] == "20-25 Years":
        List_20_25.append(survey_responses[i][5])
    
    else:
        List_above_25.append(survey_responses[i][5])


In [125]:
print(len(List_below_5) ,"people have 0-5 years of experience")
print(len(List_5_10) ,"people have 5-10 years of experience")
print(len(List_10_15) ,"people have 10-15 years of experience")
print(len(List_15_20) ,"people have 15-20 years of experience")
print(len(List_20_25) ,"people have 20-25 years of experience")
print(len(List_above_25) ,"people have +25 years of experience")


18753 people have 0-5 years of experience
3167 people have 5-10 years of experience
1118 people have 10-15 years of experience
1069 people have 15-20 years of experience
925 people have 20-25 years of experience
941 people have +25 years of experience


In [132]:
print("People with 0 - 5 years of experience have an average salary of $",round(((sum(List_below_5)/len(List_below_5))),2))
print("People with 5 - 10 years of experience have an average salary of $",round(((sum(List_5_10)/len(List_5_10))),2))
print("People with 10 - 15 years of experience have an average salary of $",round(((sum(List_10_15)/len(List_10_15))),2))
print("People with 15 - 20 years of experience have an average salary of $",round(((sum(List_15_20)/len(List_15_20))),2))
print("People with 20 - 25 years of experience have an average salary of $",round(((sum(List_20_25)/len(List_20_25))),2))
print("People with +25 years of experience have an average salary of $",round(((sum(List_above_25)/len(List_above_25))),2))


People with 0 - 5 years of experience have an average salary of $ 45047.87
People with 5 - 10 years of experience have an average salary of $ 59312.82
People with 10 - 15 years of experience have an average salary of $ 80226.76
People with 15 - 20 years of experience have an average salary of $ 75101.83
People with 20 - 25 years of experience have an average salary of $ 103159.8
People with +25 years of experience have an average salary of $ 90444.99


## 1 Question. What does the distribution of the experience categories look like? Is there an even spread of new and experienced programmers?

The distribution of the number of people, regarding programming experience is uneven. The number of people with less than 5 years of experience definitely prevails. Probably this is due to the fact that programming is a fairly young field of work.


## 2 Question. How does average compensation change as a programmer gains experience? Is there a general trend that holds? Are there interesting changes from one category to the next?

Here there is no surprise. The more experience, the higher the salary, as in many other professions. An interesting result is the average salary of those with the most experience.  Contrary to the trend, it is not the highest. 


## 3 Question. Are there any extreme values in the data? These can influence the average and mislead us on what a "typical" value is.

I did not find results significantly above average


## 4 Question. What are the most commonly used skills among survey respondents?

85% of those surveyed said they were familiar with Python. In second place is SQL with 41%.